io-engine/src/rebuild/rebuild_task.rs

```diff
@@ -110,59 +111,137 @@ impl RebuildTask {
     /// Copies one segment worth of data from source into destination.
     async fn copy_one(
         &mut self,
-        blk: u64,
-        descriptor: &RebuildDescriptor,
+        offset_blk: u64,
+        desc: &RebuildDescriptor,
     ) -> Result<(), RebuildError> {
-        let mut copy_buffer: DmaBuf;
-        let source_hdl = descriptor.src_io_handle().await?;
-        let destination_hdl = descriptor.dst_io_handle().await?;
+        if let Some(buf) = self.read_src(offset_blk, desc).await? {
+            self.write_dst(offset_blk, &buf, desc).await?;
 
-        let copy_buffer = if descriptor.get_segment_size_blks(blk)
-            == descriptor.segment_size_blks
-        {
-            &mut self.buffer
-        } else {
-            let segment_size_blks = descriptor.range.end - blk;
+            if !matches!(desc.options.verify_mode, RebuildVerifyMode::None) {
+                self.verify_dst(offset_blk, &buf, desc).await?;
+            }
+        }
 
-            debug!(
-                    "Adjusting last segment size from {} to {}. offset: {}, range: {:?}",
-                    descriptor.segment_size_blks, segment_size_blks, blk, descriptor.range,
-                );
+        Ok(())
+    }
 
-            copy_buffer = destination_hdl
-                .dma_malloc(segment_size_blks * descriptor.block_size)
-                .context(NoCopyBuffer {})?;
+    /// Reads a rebuild segment at the given offset from the source replica, and
+    /// returns the buffer to be written to the destination in the case the
+    /// segment is allocated on the source.
+    async fn read_src(
+        &self,
+        offset_blk: u64,
+        desc: &RebuildDescriptor,
+    ) -> Result<Option<DmaBuf>, RebuildError> {
+        let src = desc.src_io_handle().await?;
 
-            &mut copy_buffer
-        };
+        let num_blocks = desc.get_segment_size_blks(offset_blk);
 
-        let res = source_hdl
-            .read_at_ex(
-                blk * descriptor.block_size,
-                copy_buffer,
-                Some(ReadMode::UnwrittenFail),
-            )
-            .await;
+        let mut buf = src
```

Comment:
Why are we now allocating the buffer here? We have to allocate/deallocate every time, whilst before we ensured we had the buffers for the duration of the rebuild.

Reply:
Fixed and improved to prevent buffer allocation for the final segment.